src/algorithms/enumerate.rs

```diff
@@ -0,0 +1,94 @@
+use crate::{enumerate_mask, Correctness, Guess, Guesser, DICTIONARY, MAX_MASK_ENUM};
+use once_cell::sync::OnceCell;
+use std::borrow::Cow;
+
+static INITIAL: OnceCell<Vec<(&'static str, usize)>> = OnceCell::new();
+
+pub struct Enumerate {
+    remaining: Cow<'static, Vec<(&'static str, usize)>>,
+}
+
+impl Enumerate {
+    pub fn new() -> Self {
+        Self {
+            remaining: Cow::Borrowed(INITIAL.get_or_init(|| {
+                Vec::from_iter(DICTIONARY.lines().map(|line| {
+                    let (word, count) = line
+                        .split_once(' ')
+                        .expect("every line is word + space + frequency");
+                    let count: usize = count.parse().expect("every count is a number");
+                    (word, count)
+                }))
+            })),
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+struct Candidate {
+    word: &'static str,
+    goodness: f64,
+}
+
+impl Guesser for Enumerate {
+    fn guess(&mut self, history: &[Guess]) -> String {
+        if let Some(last) = history.last() {
+            if matches!(self.remaining, Cow::Owned(_)) {
+                self.remaining
+                    .to_mut()
+                    .retain(|(word, _)| last.matches(word));
+            } else {
+                self.remaining = Cow::Owned(
+                    self.remaining
+                        .iter()
+                        .filter(|(word, _)| last.matches(word))
+                        .copied()
+                        .collect(),
+                );
+            }
+        }
+        if history.is_empty() {
+            return "tares".to_string();
+        }
+
+        let remaining_count: usize = self.remaining.iter().map(|&(_, c)| c).sum();
+
+        let mut best: Option<Candidate> = None;
+        for &(word, count) in &*self.remaining {
+            let mut totals = [0usize; MAX_MASK_ENUM];
+            // considering a world where we _did_ guess `word` and got `pattern` as the
+            // correctness. now, compute what _then_ is left.
+            for (candidate, count) in &*self.remaining {
+                let idx = enumerate_mask(&Correctness::compute(candidate, word));
```

Comment:
Super nice observation here — basically that for any given answer candidate, there is exactly _one_ pattern that we'd get for any given guess, so there's no point looping over both patterns _and_ words. I would love to see that in a comment here though.